src/rest_framework_dso/paginator.py

```diff
@@ -0,0 +1,241 @@
+from typing import Callable, Iterable, List, Optional
+
+from django.core.paginator import EmptyPage
+from django.core.paginator import Page as DjangoPage
+from django.core.paginator import PageNotAnInteger
+from django.core.paginator import Paginator as DjangoPaginator
+from django.db.models.query import QuerySet
+from django.http import Http404
+from django.utils.functional import cached_property
+from django.utils.translation import gettext_lazy as _
+
+from rest_framework_dso.embedding import ObservableIterator
+
+
+class DSOPaginator(DjangoPaginator):
+    """A paginator that supports streaming.
+
+    This paginator avoids expensive count queries.
+    So num_pages() is not supported.
+    """
+
+    def __init__(self, object_list, per_page, orphans=0, allow_empty_first_page=True):
+        super().__init__(object_list, per_page, 0, allow_empty_first_page)
+
+    def validate_number(self, number):
+        """Validate the given 1-based page number."""
+        try:
+            if isinstance(number, float) and not number.is_integer():
+                raise ValueError
+            number = int(number)
+        except (TypeError, ValueError) as e:
+            raise PageNotAnInteger(_("That page number is not an integer")) from e
+        if number < 1:
+            raise EmptyPage(_("That page number is less than 1"))
+        return number
+
+    def get_page(self, number):
+        """
+        Return a valid page, even if the page argument isn't a number or isn't
+        in range.
+        """
+        try:
+            number = self.validate_number(number)
+        except PageNotAnInteger:
+            number = 1
+        return self.page(number)
+
+    def page(self, number):
+        """Return a Page object for the given 1-based page number."""
+        number = self.validate_number(number)
+        bottom = (number - 1) * self.per_page
+        top = bottom + self.per_page
+
+        # One additional sentinel object, is given to the page.
+        # This object should not be rendered, but it allows the page
+        # to detect whether more items exist beyond it and hence wether a next page exists.
+        sentinel = 1
+        return self._get_page(self.object_list[bottom : top + sentinel], number, self)
+
+    def _get_page(self, *args, **kwargs):
+        """
+        Return an instance of a single page.
+
+        This hook can be used by subclasses to use an alternative to the
+        standard :cls:`Page` object.
+        """
+        return DSOPage(*args, **kwargs)
+
+    @cached_property
+    def num_pages(self):
+        """Total number of pages is unknown."""
+        raise NotImplementedError()
```

Comment:
Wel handig om een message mee te geven. Zeker als we ooit in Sentry `NotImplementedError` zien zonder melding. Kost altijd meer tijd om te ontdekken wat er nou mis is.